File: src/websites/saucedemo/pages/SauceBasePage.ts
import type { Locator, Page } from '@playwright/test';
import { ElementHelper } from '@helpers/ElementHelper';

/**
 * Base for every SauceDemo page object.
 *
 * Holds locators and actions that genuinely repeat across the site —
 * primarily the header burger menu and the shopping-cart link (both
 * present once the user is authenticated). Pre-login pages still extend
 * this base for the `helper` and `page` plumbing.
 */
export abstract class SauceBasePage {
  protected readonly helper: ElementHelper;

  readonly burgerMenuButton: Locator;
  readonly logoutLink: Locator;
  readonly shoppingCartLink: Locator;
  readonly shoppingCartBadge: Locator;

  constructor(protected readonly page: Page) {
    this.helper = new ElementHelper();
    this.burgerMenuButton = page.locator('#react-burger-menu-btn');
    this.logoutLink = page.getByTestId('logout-sidebar-link');
    this.shoppingCartLink = page.getByTestId('shopping-cart-link');
    this.shoppingCartBadge = page.getByTestId('shopping-cart-badge');
  }

  async openMenu(): Promise<void> {
    await this.burgerMenuButton.click();
  }

  async logout(): Promise<void> {
    await this.openMenu();
    await this.logoutLink.click();
  }

  async openCart(): Promise<void> {
    await this.shoppingCartLink.click();
  }

  /** Returns 0 when the badge is absent (empty cart). */
  async cartBadgeCount(): Promise<number> {
    if ((await this.shoppingCartBadge.count()) === 0) return 0;
    return Number(await this.shoppingCartBadge.innerText());
  }
}

File: selenium-java-ui-demo/src/test/java/com/demo/saucedemo/pages/SauceBasePage.java
package com.demo.saucedemo.pages;

import com.demo.helpers.ElementHelper;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

/**
 * Base for every authenticated SauceDemo page: the burger menu, logout link, and cart badge are
 * present on all of them.
 *
 * Locators are {@code By} fields, not found elements — Selenium's {@code WebElement} goes stale
 * after any navigation/DOM change, so every action re-finds via {@code driver.findElement(by)}
 * at call time.
 */
public abstract class SauceBasePage {

  protected final WebDriver driver;
  protected final WebDriverWait wait;
  protected final ElementHelper helper;

  protected final By burgerMenuButton = By.id("react-burger-menu-btn");
  protected final By logoutLink = By.cssSelector("[data-test='logout-sidebar-link']");
  protected final By shoppingCartLink = By.cssSelector("[data-test='shopping-cart-link']");
  protected final By shoppingCartBadge = By.cssSelector("[data-test='shopping-cart-badge']");

  protected SauceBasePage(WebDriver driver, WebDriverWait wait) {
    this.driver = driver;
    this.wait = wait;
    this.helper = new ElementHelper(driver);
  }

  public void openMenu() {
    wait.until(ExpectedConditions.elementToBeClickable(burgerMenuButton)).click();
  }

  public void logout() {
    openMenu();
    wait.until(ExpectedConditions.elementToBeClickable(logoutLink)).click();
  }

  public void openCart() {
    driver.findElement(shoppingCartLink).click();
  }

  /** Returns 0 when the badge is absent (empty cart), instead of throwing. */
  public int cartBadgeCount() {
    var badges = driver.findElements(shoppingCartBadge);
    return badges.isEmpty() ? 0 : Integer.parseInt(badges.get(0).getText());
  }
}
